lib/hal/audio.c

```diff
@@ -55,25 +61,59 @@ static void __stdcall DPC(PKDPC Dpc,
 // PCM actions, since S/PDIF is always spooling through the same buffer.
 static BOOLEAN __stdcall ISR(PKINTERRUPT Interrupt, PVOID ServiceContext)
 {
-	// Was the interrupt triggered because we were out of data?
-	if ((*((char *)0xFEC00116))&8)
-		KeInsertQueueDpc(&DPCObject,NULL,NULL); //calls user callback soon
+	AC97_DEVICE *pac97device = &ac97Device;
+	volatile unsigned char *pb = (unsigned char *)pac97device->mmio;
 
-	//KeInsertQueueDpc queues Dpc and returns TRUE if Dpc not already queued.
-	//Dpc will be queued only once. So only one Dpc is fired after ISRs cease fire.
-	//DPCs avoid crashes inside non reentrant user callbacks called by nested ISRs.
-	//CAUTION : if you use fpu in DPC you have to save & restore yourself fpu state!!!
-	//(fpu=floating point unit, i.e the coprocessor executing floating point opcodes)
+	unsigned char analogInterrupt = pb[0x116];
+	unsigned char digitalInterrupt = pb[0x176];
+
+	bool waitCompleted = false;
+
+	if (analogInterrupt) {
+
+		bool waitingForAnalog = (analogBufferCount > digitalBufferCount);
 
-	// Was the interrupt triggered because of FIFO error?
-	if ((*((char *)0xFEC00116))&0x10)
-		{
-			// Fifo underrun - what should I do here?
-		// In case of heavy processing, insert a DPC
+		// Was the interrupt triggered because we were out of data?
+		if ((analogInterrupt & 8) && !analogDrained) {
+			if (analogBufferCount > 0) {
+				waitCompleted |= waitingForAnalog;
+				analogBufferCount--;
+			}
 		}
 
-	*((char *)0xFEC00116)=0xFF; // clear all int sources
-	*((char *)0xFEC00176)=0xFF; // clear all int sources
+		analogDrained = analogInterrupt & 4;
+
+		pb[0x116]=0xFF; // clear all int sources
+	}
+
+	if (digitalInterrupt) {
+
+		bool waitingForDigital = (digitalBufferCount > analogBufferCount);
+
+		// Was the interrupt triggered because we were out of data?
+		if ((digitalInterrupt & 8) && !digitalDrained) {
+			if (digitalBufferCount > 0) {
+				waitCompleted |= waitingForDigital;
+				digitalBufferCount--;
+			}
+		}
+
+		digitalDrained = digitalInterrupt & 4;
+
+		pb[0x176]=0xFF; // clear all int sources
+	}
+
+
+	// If a buffer was consumed by analog and digital output, we ask for a DPC
+	if (waitCompleted) {
+			KeInsertQueueDpc(&DPCObject,NULL,NULL); //calls user callback soon
```

Comment:
nit: Indentation

Reply:
This came from the original code. I've re-arranged it, and added another commit to also improve the `DPC` function, which had similar issues.

It's still not ideal, but better. The goal of this PR is functionality, not style.